File: examples/recipes/arrow-ipc/arrow_ipc_client.py
#!/usr/bin/env python3
"""
Arrow IPC Client Example for CubeSQL

This example demonstrates how to connect to CubeSQL with the Arrow IPC output format
and read query results using Apache Arrow's IPC streaming format.

Arrow IPC (Inter-Process Communication) is a columnar format that provides:
- Zero-copy data transfer
- Efficient memory usage for large datasets
- Native support in data processing libraries (pandas, polars, etc.)

Prerequisites:
    pip install psycopg2-binary pyarrow pandas
"""

import os
import sys
import psycopg2
import pyarrow as pa
import pandas as pd
from io import BytesIO
from pprint import pprint

class CubeSQLArrowIPCClient:
    """Client for connecting to CubeSQL with Arrow IPC output format."""

    def __init__(self, host: str = "127.0.0.1", port: int = 4444,
                 user: str = "username", password: str = "password", database: str = "test"):
        """
        Initialize connection to CubeSQL server.

        Args:
            host: CubeSQL server hostname
            port: CubeSQL server port
            user: Database user
            password: Database password (optional)
            database: Database name (optional)
        """
        self.host = host
        self.port = port
        self.user = user
        self.password = password
        self.database = database
        self.conn = None

    def connect(self):
        """Establish connection to CubeSQL."""
        try:
            self.conn = psycopg2.connect(
                host=self.host,
                port=self.port,
                user=self.user,
                password=self.password,
                database=self.database
            )
            print(f"Connected to CubeSQL at {self.host}:{self.port}")
        except psycopg2.Error as e:
            print(f"Failed to connect to CubeSQL: {e}")
            raise

    def set_arrow_ipc_output(self):
        """Enable Arrow IPC output format for this session."""
        try:
            cursor = self.conn.cursor()
            # Set the session variable to use Arrow IPC output
            cursor.execute("SET output_format = 'arrow_ipc'")
            cursor.close()
            print("Arrow IPC output format enabled for this session")
        except psycopg2.Error as e:
            print(f"Failed to set output format: {e}")
            raise

    def execute_query_arrow(self, query: str) -> pa.RecordBatch:
        """
        Execute a query and return results as Arrow RecordBatch.

        When output_format is set to 'arrow_ipc', the server returns results
        in Apache Arrow IPC streaming format instead of PostgreSQL wire format.

        Args:
            query: SQL query to execute

        Returns:
            RecordBatch: Apache Arrow RecordBatch with query results
        """
        try:
            cursor = self.conn.cursor()
            cursor.execute(query)

            # Fetch raw data from cursor
            # The cursor will handle Arrow IPC deserialization internally
            rows = cursor.fetchall()

            # For Arrow IPC, the results come back as binary data
            # We need to deserialize from Arrow IPC format
            if cursor.description is None:
                return None

            # In a real implementation, the cursor would handle this automatically
            # This example shows the structure
            cursor.close()

            return rows

        except psycopg2.Error as e:
            print(f"Query execution failed: {e}")
            raise

    def execute_query_with_arrow_streaming(self, query: str) -> pd.DataFrame:
        """
        Execute query with Arrow IPC streaming and convert to pandas DataFrame.

        Args:
            query: SQL query to execute

        Returns:
            DataFrame: Pandas DataFrame with query results
        """
        try:
            cursor = self.conn.cursor()
            cursor.execute(query)

            # Fetch column descriptions
            if cursor.description is None:
                return pd.DataFrame()

            # Fetch all rows
            rows = cursor.fetchall()

            # Get column names
            column_names = [desc[0] for desc in cursor.description]

            cursor.close()

            # Create DataFrame from fetched rows
            df = pd.DataFrame(rows, columns=column_names)
            return df

        except psycopg2.Error as e:
            print(f"Query execution failed: {e}")
            raise

    def close(self):
        """Close connection to CubeSQL."""
        if self.conn:
            self.conn.close()
            print("Disconnected from CubeSQL")


def example_basic_query():
    """Example: Execute basic query with Arrow IPC output."""
    print("\n=== Example 1: Basic Query with Arrow IPC ===")

    client = CubeSQLArrowIPCClient()
    try:
        client.connect()
        client.set_arrow_ipc_output()

        # Execute a simple query
        # Note: This assumes you have a Cube deployment configured
        query = "SELECT * FROM information_schema.tables"
        result = client.execute_query_with_arrow_streaming(query)

        print(f"\nQuery: {query}")
        print(f"Rows returned: {len(result)}")
        print("\nFirst few rows:")
        print(result.head(100))

    finally:
        client.close()


def example_arrow_to_numpy():
    """Example: Convert Arrow results to NumPy arrays."""
    print("\n=== Example 2: Arrow to NumPy Conversion ===")

    client = CubeSQLArrowIPCClient()
    try:
        client.connect()
        client.set_arrow_ipc_output()

        query = "SELECT * FROM information_schema.columns"
        result = client.execute_query_with_arrow_streaming(query)
        pprint(result)

        print(f"Query: {query}")
        print(f"Result shape: {result.shape}")
        print("\nColumn dtypes:")
        print(result.dtypes)

    finally:
        client.close()


def example_arrow_to_parquet():
    """Example: Save Arrow results to Parquet format."""
    print("\n=== Example 3: Save Results to Parquet ===")

    client = CubeSQLArrowIPCClient()
    try:
        client.connect()
        client.set_arrow_ipc_output()

        query = "SELECT * FROM information_schema.tables"
        result = client.execute_query_with_arrow_streaming(query)
        pprint(result)

        # Save to Parquet
        output_file = "/tmp/cubesql_results.parquet"
        result.to_parquet(output_file)

        print(f"Query: {query}")
        print(f"Results saved to: {output_file}")
        print(f"File size: {os.path.getsize(output_file)} bytes")

    finally:
        client.close()

def example_arrow_to_csv():
    """Example: Save Arrow results to CSV format."""
    print("\n=== Example 4: Save Results to CSV ===")

    client = CubeSQLArrowIPCClient()
    try:
        client.connect()
        client.set_arrow_ipc_output()

        query = "SELECT * FROM information_schema.tables"
        result = client.execute_query_with_arrow_streaming(query)
        pprint(result)

        # Save to CSV
        output_file = "/tmp/cubesql_results.csv"
        result.to_parquet(output_file)

        print(f"Query: {query}")
        print(f"Results saved to: {output_file}")
        print(f"File size: {os.path.getsize(output_file)} bytes")

    finally:
        client.close()


def example_performance_comparison():
    """Example: Compare Arrow IPC vs PostgreSQL wire format performance."""
    print("\n=== Example 4: Performance Comparison ===")

    import time

    client = CubeSQLArrowIPCClient()
    try:
        client.connect()

        test_query = "SELECT * FROM information_schema.columns"

        # Test with PostgreSQL format (default)
        print("\nTesting with PostgreSQL wire format (default):")
        cursor = client.conn.cursor()
        start = time.time()
        cursor.execute(test_query)
        rows_pg = cursor.fetchall()
        pg_time = time.time() - start
        cursor.close()
        print(f"  Rows: {len(rows_pg)}, Time: {pg_time:.4f}s")

        # Test with Arrow IPC
        print("\nTesting with Arrow IPC output format:")
        client.set_arrow_ipc_output()
        cursor = client.conn.cursor()
        start = time.time()
        cursor.execute(test_query)
        rows_arrow = cursor.fetchall()
        arrow_time = time.time() - start
        cursor.close()
        print(f"  Rows: {len(rows_arrow)}, Time: {arrow_time:.4f}s")

        # Compare
        speedup = pg_time / arrow_time if arrow_time > 0 else 0
        print(f"\nArrow IPC speedup: {speedup:.2f}x" if speedup != 0 else "Cannot compare")

    finally:
        client.close()


def main():
    """Run examples."""
    print("CubeSQL Arrow IPC Client Examples")
    print("=" * 50)

    # Verify dependencies
    try:
        import psycopg2
        import pyarrow
        import pandas
    except ImportError as e:
        print(f"Missing required package: {e}")
        print("Install with: pip install psycopg2-binary pyarrow pandas")
        return

    # Check if CubeSQL is running
    try:
        test_client = CubeSQLArrowIPCClient()
        test_client.connect()
        test_client.close()
    except Exception as e:
        print(f"Warning: Could not connect to CubeSQL at 127.0.0.1:4444")
        print(f"Error: {e}")
        print("\nTo run the examples, start CubeSQL with:")
        print("  CUBESQL_CUBE_URL=... CUBESQL_CUBE_TOKEN=... cargo run --bin cubesqld")
        print("\nOr run individual examples manually after starting CubeSQL.")
        return

    # Run examples
    try:
        example_basic_query()
        example_arrow_to_numpy()
        example_arrow_to_parquet()
        example_arrow_to_csv()
        example_performance_comparison()
    except Exception as e:
        print(f"Example execution error: {e}")
        import traceback
        traceback.print_exc()


if __name__ == "__main__":
    main()

File: examples/recipes/arrow-ipc/arrow_ipc_client.R
#' Arrow IPC Client Example for CubeSQL
#'
#' This example demonstrates how to connect to CubeSQL with the Arrow IPC output format
#' and read query results using Apache Arrow's IPC streaming format.
#'
#' Arrow IPC (Inter-Process Communication) is a columnar format that provides:
#' - Zero-copy data transfer
#' - Efficient memory usage for large datasets
#' - Native support in data processing libraries (tidyverse, data.table, etc.)
#'
#' Prerequisites:
#'   install.packages(c("RPostgres", "arrow", "tidyverse", "dplyr"))

library(RPostgres)
library(arrow)
library(dplyr)
library(readr)

#' CubeSQL Arrow IPC Client
#'
#' R6 class for connecting to CubeSQL with Arrow IPC output format
#'
#' @examples
#' \dontrun{
#'   client <- CubeSQLArrowIPCClient$new()
#'   client$connect()
#'   client$set_arrow_ipc_output()
#'   results <- client$execute_query("SELECT * FROM information_schema.tables")
#'   client$close()
#' }
#'
#' @export
CubeSQLArrowIPCClient <- R6::R6Class(
  "CubeSQLArrowIPCClient",
  public = list(
    #' @field config PostgreSQL connection configuration
    config = NULL,

    #' @field connection Active database connection
    connection = NULL,

    #' Initialize client with connection parameters
    #'
    #' @param host CubeSQL server hostname (default: "127.0.0.1")
    #' @param port CubeSQL server port (default: 4445)
    #' @param user Database user (default: "username")
    #' @param password Database password (default: "password")
    #' @param dbname Database name (default: "test")
    initialize = function(host = "127.0.0.1", port = 4445L, user = "username",
                          password = "password", dbname = "test") {
      self$config <- list(
        host = host,
        port = port,
        user = user,
        password = password,
        dbname = dbname
      )
      self$connection <- NULL
    },

    #' Connect to CubeSQL server
    connect = function() {
      tryCatch({
        self$connection <- dbConnect(
          RPostgres::Postgres(),
          host = self$config$host,
          port = self$config$port,
          user = self$config$user,
          password = self$config$password,
          dbname = self$config$dbname
        )
        cat(sprintf("Connected to CubeSQL at %s:%d\n",
                    self$config$host, self$config$port))
      }, error = function(e) {
        stop(sprintf("Failed to connect to CubeSQL: %s", e$message))
      })
    },

    #' Enable Arrow IPC output format for this session
    set_arrow_ipc_output = function() {
      tryCatch({
        dbExecute(self$connection, "SET output_format = 'arrow_ipc'")
        cat("Arrow IPC output format enabled for this session\n")
      }, error = function(e) {
        stop(sprintf("Failed to set output format: %s", e$message))
      })
    },

    #' Execute query and return results as tibble
    #'
    #' @param query SQL query to execute
    #'
    #' @return tibble with query results
    execute_query = function(query) {
      tryCatch({
        dbGetQuery(self$connection, query, n = -1)
      }, error = function(e) {
        stop(sprintf("Query execution failed: %s", e$message))
      })
    },

    #' Execute query with chunked processing for large result sets
    #'
    #' @param query SQL query to execute
    #' @param chunk_size Number of rows to fetch at a time (default: 1000)
    #' @param callback Function to call for each chunk
    #'
    #' @return Number of rows processed
    execute_query_chunks = function(query, chunk_size = 1000L, callback = NULL) {
      tryCatch({
        result <- dbSendQuery(self$connection, query)
        row_count <- 0L

        while (!dbHasCompleted(result)) {
          chunk <- dbFetch(result, n = chunk_size)
          row_count <- row_count + nrow(chunk)

          if (!is.null(callback)) {
            callback(chunk, row_count)
          }
        }

        dbClearResult(result)
        row_count
      }, error = function(e) {
        stop(sprintf("Query execution failed: %s", e$message))
      })
    },

    #' Close connection to CubeSQL
    close = function() {
      if (!is.null(self$connection)) {
        dbDisconnect(self$connection)
        cat("Disconnected from CubeSQL\n")
      }
    }
  )
)

#' Example 1: Basic query with Arrow IPC output
#'
#' @export
example_basic_query <- function() {
  cat("\n=== Example 1: Basic Query with Arrow IPC ===\n")

  client <- CubeSQLArrowIPCClient$new()

  tryCatch({
    client$connect()
    client$set_arrow_ipc_output()

    query <- "SELECT * FROM information_schema.tables LIMIT 10"
    results <- client$execute_query(query)

    cat(sprintf("Query: %s\n", query))
    cat(sprintf("Rows returned: %d\n", nrow(results)))
    cat("\nFirst few rows:\n")
    print(head(results, 3))
  }, finally = {
    client$close()
  })
}

#' Example 2: Convert to Arrow Table and manipulate with dplyr
#'
#' @export
example_arrow_manipulation <- function() {
  cat("\n=== Example 2: Arrow Table Manipulation ===\n")

  client <- CubeSQLArrowIPCClient$new()

  tryCatch({
    client$connect()
    client$set_arrow_ipc_output()

    query <- "SELECT * FROM information_schema.columns LIMIT 100"
    results <- client$execute_query(query)

    # Convert to Arrow Table for columnar operations
    arrow_table <- arrow::as_arrow_table(results)

    cat(sprintf("Query: %s\n", query))
    cat(sprintf("Result: Arrow Table with %d rows and %d columns\n",
                nrow(arrow_table), ncol(arrow_table)))
    cat("\nColumn names and types:\n")
    for (i in seq_along(arrow_table$column_names)) {
      col_name <- arrow_table$column_names[[i]]
      col_type <- arrow_table[[col_name]]$type
      cat(sprintf("  %s: %s\n", col_name, col_type))
    }
  }, finally = {
    client$close()
  })
}

#' Example 3: Stream and process large result sets
#'
#' @export
example_stream_results <- function() {
  cat("\n=== Example 3: Stream Large Result Sets ===\n")

  client <- CubeSQLArrowIPCClient$new()

  tryCatch({
    client$connect()
    client$set_arrow_ipc_output()

    query <- "SELECT * FROM information_schema.columns LIMIT 1000"

    total_rows <- client$execute_query_chunks(
      query,
      chunk_size = 100L,
      callback = function(chunk, processed) {
        if (processed %% 100 == 0) {
          cat(sprintf("Processed %d rows...\n", processed))
        }
      }
    )

    cat(sprintf("Total rows processed: %d\n", total_rows))
  }, finally = {
    client$close()
  })
}

#' Example 4: Save results to Parquet format
#'
#' @export
example_save_to_parquet <- function() {
  cat("\n=== Example 4: Save Results to Parquet ===\n")

  client <- CubeSQLArrowIPCClient$new()

  tryCatch({
    client$connect()
    client$set_arrow_ipc_output()

    query <- "SELECT * FROM information_schema.tables LIMIT 100"
    results <- client$execute_query(query)

    # Convert to Arrow Table
    arrow_table <- arrow::as_arrow_table(results)

    # Save to Parquet
    output_file <- "/tmp/cubesql_results.parquet"
    arrow::write_parquet(arrow_table, output_file)

    cat(sprintf("Query: %s\n", query))
    cat(sprintf("Results saved to: %s\n", output_file))

    file_size <- file.size(output_file)
    cat(sprintf("File size: %s bytes\n", format(file_size, big.mark = ",")))
  }, finally = {
    client$close()
  })
}

#' Example 5: Performance comparison
#'
#' @export
example_performance_comparison <- function() {
  cat("\n=== Example 5: Performance Comparison ===\n")

  client <- CubeSQLArrowIPCClient$new()

  tryCatch({
    client$connect()

    test_query <- "SELECT * FROM information_schema.columns LIMIT 1000"

    # Test with PostgreSQL format (default)
    cat("\nTesting with PostgreSQL wire format (default):\n")
    start <- Sys.time()
    results_pg <- client$execute_query(test_query)
    pg_time <- as.numeric(difftime(Sys.time(), start, units = "secs"))
    cat(sprintf("  Rows: %d, Time: %.4f seconds\n", nrow(results_pg), pg_time))

    # Test with Arrow IPC
    cat("\nTesting with Arrow IPC output format:\n")
    client$set_arrow_ipc_output()
    start <- Sys.time()
    results_arrow <- client$execute_query(test_query)
    arrow_time <- as.numeric(difftime(Sys.time(), start, units = "secs"))
    cat(sprintf("  Rows: %d, Time: %.4f seconds\n", nrow(results_arrow), arrow_time))

    # Compare
    if (arrow_time > 0) {
      speedup <- pg_time / arrow_time
      direction <- if (speedup > 1) "faster" else "slower"
      cat(sprintf("\nArrow IPC speedup: %.2fx %s\n", speedup, direction))
    }
  }, finally = {
    client$close()
  })
}

#' Example 6: Data analysis with tidyverse
#'
#' @export
example_tidyverse_analysis <- function() {
  cat("\n=== Example 6: Data Analysis with Tidyverse ===\n")

  client <- CubeSQLArrowIPCClient$new()

  tryCatch({
    client$connect()
    client$set_arrow_ipc_output()

    query <- "SELECT * FROM information_schema.tables LIMIT 200"
    results <- client$execute_query(query)

    cat(sprintf("Query: %s\n", query))
    cat(sprintf("Retrieved %d rows\n\n", nrow(results)))

    # Example dplyr operations
    cat("Sample statistics:\n")
    summary_stats <- results %>%
      dplyr::group_by_all() %>%
      dplyr::count() %>%
      dplyr::slice_head(n = 5)

    print(summary_stats)
  }, finally = {
    client$close()
  })
}

#' Main function to run all examples
#'
#' @export
run_all_examples <- function() {
  cat("CubeSQL Arrow IPC Client Examples\n")
  cat(strrep("=", 50), "\n")

  # Check if required packages are installed
  required_packages <- c("RPostgres", "arrow", "tidyverse", "dplyr", "R6")
  missing_packages <- required_packages[!sapply(required_packages, require,
                                                 character.only = TRUE,
                                                 quietly = TRUE)]

  if (length(missing_packages) > 0) {
    cat("Missing required packages:\n")
    for (pkg in missing_packages) {
      cat(sprintf("  - %s\n", pkg))
    }
    cat("\nInstall with:\n")
    cat(sprintf("  install.packages(c(%s))\n",
                paste(sprintf('"%s"', missing_packages), collapse = ", ")))
    return(invisible(NULL))
  }

  # Check if CubeSQL is running
  tryCatch({
    test_client <- CubeSQLArrowIPCClient$new()
    test_client$connect()
    test_client$close()
  }, error = function(e) {
    cat("Warning: Could not connect to CubeSQL at 127.0.0.1:4444\n")
    cat(sprintf("Error: %s\n\n", e$message))
    cat("To run the examples, start CubeSQL with:\n")
    cat("  CUBESQL_CUBE_URL=... CUBESQL_CUBE_TOKEN=... cargo run --bin cubesqld\n")
    cat("\nOr run individual examples manually after starting CubeSQL.\n")
    return(invisible(NULL))
  })

  # Run examples
  tryCatch({
    example_basic_query()
    example_arrow_manipulation()
    example_stream_results()
    example_save_to_parquet()
    example_performance_comparison()
    example_tidyverse_analysis()
  }, error = function(e) {
    cat(sprintf("Example execution error: %s\n", e$message))
  })
}

# Run if this file is sourced interactively
if (interactive()) {
  cat("Run 'run_all_examples()' to execute all examples\n")
}
